docs/caching_and_memoization.rst

```diff
@@ -0,0 +1,316 @@
+-------------------------
+
+Cached functions usually have a set of parameters passed in, and will return
+different results for different sets of parameters.
+
+Best practice here is to use as small a set of parameters as possible, and to
+use simple objects as parameters when possible (strings, booleans, integers,
+that sort of thing).
+
+.. code-block:: python
+
+    @quickcache(['domain_obj.name', 'user._id'], timeout=10)
+    def count_users_forms_by_device(domain_obj, user):
+        return {
+            FormAccessors(domain_obj.name).count_forms_by_device(device.device_id)
+            for device in user.devices
+        }
+
+The first argument to ``@quickcache`` is an argument called ``vary_on`` which is
+a list of the parameters used to identify each result stored in the cache. Taken
+together, the variables specified in vary_on should constitute all inputs that
+would change the value of the output. You may be thinking “Well, isn’t that just
+all of the arguments?” Often, yes. However, also very frequently, a function
+depends not on the exact object being passed in, but merely on one or a few
+properties of that object. In the example above, we want the function to return
+the same result when called with the same domain name and user ID, not
+necessarily the same exact objects. Quickcache handles this by allowing you to
+pass in strings like ``parameter.attribute``. Additionally, instead of a list of
+parameters, you may pass in a function, which will be called with the arguments
+of the cached function to return a cache key.
+
+Memoized does not provide these capabilities, and instead always uses all of the
+arguments passed in. For this reason, you should only memoize functions with
+simple arguments. At a minimum, all arguments to memoized must be hashable.
+You'll notice that the above function doesn't actually use anything on the
+``domain_obj`` other than name, so you could just refactor it to accept
+``domain`` instead (this also means code calling this function won't need to
+fetch the domain object to pass to this function, only to discard everything
+except the name anyways).
+
+You don't need to let this consideration muck up your function's interface. A
+common practice is to make a helper function with simple arguments, and decorate
+that. You can then still use the top-level function as you see fit. For example,
+let's pretend the above function is an instance method and you want to use
+memoize rather than quickcache. You could split it apart like this:
+
+.. code-block:: python
+
+    @memoized
+    def _count_users_forms_by_device(self, domain, device_ids):
+        return {
+            FormAccessors(domain.name).count_forms_by_device(device_id)
+            for device_id in device_ids
+        }
+
+    def count_users_forms_by_device(self, domain_obj, user):
+        return self._count_users_forms_by_device(
+            domain_obj.name,
+            [device.device_id for device in user.devices]
```

Comment:
will this work with memoized (lists are not hashable)? I think you might need to use something like `frozenset(device.device_id for device in user.devices)`